Spektrum SRXL GPS telemetry frames added

robertgosz opened the pull request:
This change enhances the current Spektrum SRXL telemetry implementation by adding two additional frames containing GPS information. This allows viewing and logging GPS driven data (distance, speed, altitude, position...) in Spektrum telemetry enabled radios. 

![gps_screen_public](https://user-images.githubusercontent.com/10808082/50724966-786e3b00-10f6-11e9-98fd-400b11537a6a.jpg)
![gps_data_public](https://user-images.githubusercontent.com/10808082/50724969-7ad09500-10f6-11e9-90d1-b9d91585e74c.jpg)

Additional notes:
- the flag: #define GPS_INFO_FLAGS_GPS_DATA_RECEIVED_BIT is not really implemented at the moment and considered to be always true. This might be improved. 
- the change is aligned with refactoring made inside gps.h for 4.0
- it was not yet tested with 4.0 apart from compilation and flashing
- a bit of testing was successfully done under 3.5 branch 

Any help with improving and/or testing this change appreciated.

AndersHoglund commented on src/main/telemetry/srxl.c:
```diff
@@ -160,6 +162,168 @@ bool srxlFrameRpm(sbuf_t *dst, timeUs_t currentTimeUs)
+    sbufWriteU8(dst, hdopBcd);
+    sbufWriteU8(dst, gpsFlags);
+
+    return true;
```
This will cause GPS tm frames to be sent unconditionally even if there are no GPS data available and even if no GPS unit is connected at all. You should check for available data and only then `return true` else `return false` to avoid wasting precious tm bandwidth. Compare how `srlFrameVTX()` and  `srxlFrameFlightPackCurrent()` is done.

AndersHoglund commented on src/main/telemetry/srxl.c:
```diff
@@ -160,6 +162,168 @@ bool srxlFrameRpm(sbuf_t *dst, timeUs_t currentTimeUs)
+    sbufWriteU8(dst, altitudeHighBcd);
+    sbufFill(dst, 0xFF, STRU_TELE_GPS_STAT_EMPTY_FIELDS_COUNT);
+
+    return true;
```
Same as above comment.

robertgosz replied:
I have a few ideas about how to solve that. The simplest one is: assuming that (by it's core nature) GPS data changes are quite frequent we could just send GPS frames when GPS fix is acquired and there are at least 6 sats available (Spektrum TX does not display GPS data when there are less than 6 sats). Return false otherwise. Simple, fast and easy to do.

`
if (!STATE(GPS_FIX) || gpsSol.numSat < 6) { return false; }
`

Could that be qualified as an acceptable solution ?

etracer65 replied:
Also you need to check if the feature is enabled by the user. It's good that the code is bounded by `#ifdef USE_GPS`, but that only determines whether GPS code is compiled in - not whether the user has actually connected and enabled a GPS receiver. You should also check `featureIsEnabled(FEATURE_GPS)` before committing to sending these telemetry frames. Note that this has nothing to do with whether the GPS is actually configured properly, is working and has a good fix - just that the user has at least enabled the GPS feature.

AndersHoglund commented:
Agree with @etracer65 , looks good basically.  However,here are also some cosmetics like a few lines with trailing white spaces. 
Also some magic numbers used. There are symbolic defines for those, even though commented out. De-comment and use them.

stawiski commented on src/main/telemetry/srxl.c:
```diff
@@ -160,6 +162,168 @@ bool srxlFrameRpm(sbuf_t *dst, timeUs_t currentTimeUs)
+
+uint32_t dec2bcd_r(uint16_t dec)
+{
+    return (dec) ? ((dec2bcd_r( dec / 10 ) << 4) + (dec % 10)) : 0;
```
This is a recursive function (from stackoverflow?), and I'd advise against this and rewrite it into a loop.

robertgosz replied:
Yes, you are absolutely right. It was working, it was short so I used it. Unfortunately I don't know how to measure the real performance of both approaches on ARM, but I have written a simple iterative equivalent:

```
static uint32_t dec2bcd(uint16_t dec)
{
    uint32_t result = 0;
    uint8_t counter = 0;

    while (dec % 10 > 0) {
        result |= (dec % 10) << counter * 4;
        counter++;
        dec /= 10;
    }
    return result;
}
```

I also made it static. Does it look like a satisfying solution ?

stawiski replied:
I've written a little python script for you that implements the code above:
```python
import numpy as np

def dec2bcd(dec):
    dec = np.uint16(dec)
    result = np.uint32(0)
    counter = np.uint8(0)
    while ((dec % 10) > 0):
        result |= (dec % 10) << counter * 4
        counter += 1
        dec = np.uint16(dec / 10)
    return np.uint32(result)

test_dec = np.uint16(0)

while (test_dec <= 30):
    test_bcd = dec2bcd(test_dec)
    print("dec2bcd({:d}) = {:d}".format(test_dec, test_bcd))
    test_dec += 1
```

Looking at the output one can notice that this doesn't work for multiples of 10:
```
dec2bcd(0) = 0
dec2bcd(1) = 1
dec2bcd(2) = 2
dec2bcd(3) = 3
dec2bcd(4) = 4
dec2bcd(5) = 5
dec2bcd(6) = 6
dec2bcd(7) = 7
dec2bcd(8) = 8
dec2bcd(9) = 9
dec2bcd(10) = 0
dec2bcd(11) = 17
dec2bcd(12) = 18
dec2bcd(13) = 19
dec2bcd(14) = 20
dec2bcd(15) = 21
dec2bcd(16) = 22
dec2bcd(17) = 23
dec2bcd(18) = 24
dec2bcd(19) = 25
dec2bcd(20) = 0
```

stawiski replied:
You need a do { } while instead like this:
```python
import numpy as np

def dec2bcd(dec):
    dec = np.uint16(dec)
    result = np.uint32(0)
    counter = np.uint8(0)
    while True:
        result |= (dec % 10) << counter * 4
        counter += 1
        dec = np.uint16(dec / 10)
        if not ((dec % 10) > 0):
            break
    return np.uint32(result)

test_dec = np.uint16(0)

while (test_dec <= 30):
    test_bcd = dec2bcd(test_dec)
    print("dec2bcd({:d}) = {:d}".format(test_dec, test_bcd))
    test_dec += 1
```

robertgosz replied:
Shame on me. That's an obvious mistake. I should be beaten with a stick for not testing this carefully. Thanks for pointing this out. I think it should be enough to just replace `while (dec % 10 > 0)` with `while (dec)`. As I don't have access to any other programming stuff at the moment I have tested it as a Java method and it passes tests for all integers between 0 and 65535. Will check again with C version later today.

```java
    private static int dec2bcd (int dec) {
        int counter = 0;
        int result  = 0;
        
        while (dec > 0) {
            result |= (dec % 10) << counter * 4;
            counter++;
            dec /= 10;
        }
        return result;
    }

```

robertgosz commented:
I have updated the PR according to your suggestions. Also fixed some minor obvious things I found myself (like the funny altitude lo-bits calculation that just needed single modulo). Hope it looks a bit better now.

AndersHoglund commented:
Much better, thanks. No more empty GPS TM frames with no GPS is connected. I can only do this kind of negative regression testing, I have no GPS unit. But it is looking good to me.

etracer65 commented on src/main/telemetry/srxl.c:
```diff
@@ -160,6 +162,180 @@ bool srxlFrameRpm(sbuf_t *dst, timeUs_t currentTimeUs)
+    hdopBcd = dec2bcd(hdop);
+
+    // flags
+    gpsFlags = (gpsSol.llh.lat > 0)   ? (0x00     | GPS_FLAGS_IS_NORTH_BIT) : gpsFlags;
```
Looks like there's a problem here. If the condition is false than an uninitialized `gpsFlags` value will be used.

I would recommend moving the section below
```
    gpsFlags = gpsFlags | GPS_FLAGS_GPS_DATA_RECEIVED_BIT;
    gpsFlags = gpsFlags | GPS_FLAGS_GPS_FIX_VALID_BIT;
    gpsFlags = gpsFlags | GPS_FLAGS_3D_FIX_BIT;
```
above the conditionals and use it as an initializer and also consolidate into a single line. Additionally the flag setting statements can be simplified.

So it would look like:
```
    // flags
    gpsFlags = GPS_FLAGS_GPS_DATA_RECEIVED_BIT | GPS_FLAGS_GPS_FIX_VALID_BIT | GPS_FLAGS_3D_FIX_BIT;
    gpsFlags |= (gpsSol.llh.lat > 0) ? GPS_FLAGS_IS_NORTH_BIT : 0;
    gpsFlags |= (gpsSol.llh.lon > 0) ? GPS_FLAGS_IS_EAST_BIT : 0;
    gpsFlags |= (gpsSol.llh.altCm < 0) ? GPS_FLAGS_NEGATIVE_ALT_BIT : 0;
    gpsFlags |= (gpsSol.llh.lon / GPS_DEGREES_DIVIDER > 99) ? GPS_FLAGS_LONGITUDE_GREATER_99_BIT : 0;
```

robertgosz replied:
That's perfect!

etracer65 commented on src/main/telemetry/srxl.c:
```diff
@@ -160,6 +162,180 @@ bool srxlFrameRpm(sbuf_t *dst, timeUs_t currentTimeUs)
+    speedKnotsBcd = (speedTmp > 9999) ? dec2bcd(9999) : dec2bcd(speedTmp);
+
+    // RTC
+    #ifdef USE_RTC_TIME
```
Per the coding guidelines compiler directives should be left justified.

robertgosz replied:
Sure, forgot about this one.

robertgosz commented:
PR updated. I must admit you are all on top of this stuff and have a good eye :D Thanks for all your help!

rolial commented:
@robertgosz How can I get the GPS information on the IX12 radio? I'm using the 4649T rx with FC Matek F405-CTR.
I know the code is in development but could I test it before leaving the final version?
Could you pass me the step by step for me to insert in the Cli of FC Matek F405-CTR?

mikeller commented on src/main/telemetry/srxl.c:
```diff
@@ -160,6 +162,174 @@ bool srxlFrameRpm(sbuf_t *dst, timeUs_t currentTimeUs)
+static void GPStoDDDMM_MMMM(int32_t mwiigps, gpsCoordinateDDDMMmmmm_t *result)
+{
+    int32_t absgps, deg, min;
+    absgps = ABS(mwiigps);
```
Please do not align the `=`. If done, this will result in unneeded change whenever a new assignment to a variable with a longer name is added.

robertgosz replied:
This is actually not my code, but a function taken directly from the existing frsky_hub.c
Of course, I don't see any problem with changing it.

mikeller commented on src/main/telemetry/srxl.c:
```diff
@@ -160,6 +162,174 @@ bool srxlFrameRpm(sbuf_t *dst, timeUs_t currentTimeUs)
+        timeProvided = true;
+    }
+#endif
+    timeBcd = (timeProvided) ? timeBcd : 0xFFFFFFFF;
```
Why are you sending `0xFFFFFFFF` here when there is no RTC time available? Is this a special code for 'no time'? If so, using a define like `SPEKTRUM_TIME_UNKNOWN` for it instead of having a 'magic number' in code is a nice way of making this more obvious.

robertgosz replied:
Good idea!
